hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+| Inserting a new key and value | O(1)         | O(n)       |
+
+### !callout-info
+
+## We Describe Hash Table Big O by Average Case
+
+Normally, when we estimate the runtime of a data structure's methods with Big O, we assume the worst case. However, a well-designed hash table will attempt to prevent the worst-case scenario from occurring. Therefore when we work with hash tables we **assume the average-case**. The likelihood of the worst case from occurring depends on how well the **hash function** is implemented.
+
+### !end-callout
+
+## Hash Function
+
+**Hash functions** are functions responsible for mapping a given key into an integer index in the internal array.
+
+### Example Hash Function
+
+Imagine that we have a hash table which holds "name"-"job title" key-value pairs. This hash table needs a hash function to transform the "name" into an index.
+
+An example hash function could use the following algorithm:
+
+- Convert the key to a string
+- Grab the length of this string
+- Get modulo of the length of the string and the length of the internal array
+- Use this modulo as the index
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+### !callout-info
+
+## Why Modulo in This Example?
+
+When arrays get created, they get created with a fixed size in memory. This means that they have a maximum length and _can't_ hold an infinite number of things (unless there is an operation which expands the array. Python's list method `append` manages this for us!). Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!
+
+### !end-callout
+
+## Collisions
+
+There is an enormous number of possible keys and a limited-size array. At some point, a hash function will result in some keys being mapped to the same index. When this occurs, it is called a _collision_.
+
+### Example Collision
+
+Let's use the same hash function from before:
+
+```python
+def hash_code(key):
+  return len(str(key)) % len(hash_array)
+```
+
+Imagine a hash table that stores "name"-"job title" key-value pairs. The hash table's internal `hash_array` has a length of `15`. Imagine inserting the following pairs, which have the following indices after applying the above hash function:
+
+| Key (Name) | Value (Job Title)     | Index (`hash_code(key)`) |
+| ---------- | --------------------- | ------------------------ |
+| `"dan"`    | `"Lead Instructor"`   | `3`                      |
+| `"chris"`  | `"CS Fun Instructor"` | `5`                      |
+| `"dee"`    | `"Lead Instructor"`   | `3`                      |
+
+If we insert Dan first, and then eventually Dee, our hash function will attempt to add Dee at the index `3`, even though Dan already exists there.
+
+### Resolving Collisions
+
+There are many ways to resolve a collision! One way is to make each element of the hash array a collection itself, and then store new elements into the collection.
```

Comment:
distinguish between the plain internal array and the extra collection
```suggestion
There are many ways to resolve a collision! One way is to make each element of the hash array a collection itself, and then store new elements in the collection rather than directly in the hash array.
```